### PR TITLE
i2d_ASN1_OBJECT(): allocate memory if the user didn't provide a buffer

### LINKED ISSUE #6914: i2d_ASN1_OBJECT issue
Hello!

I have a question about i2d_ASN_OBJECT function. The [documentation](https://www.openssl.org/docs/man1.1.0/crypto/i2d_X509.html) says that:
> If *ppout is NULL memory will be allocated for a buffer and the encoded data written to it. In this case *ppout is not incremented and it points to the start of the data just written.
The function is declared as:
```c
int i2d_ASN1_OBJECT(const ASN1_OBJECT *a, unsigned char **pp)
```
So there is no need for manual memory allocation for the "pp"?
But I am getting segfault in this code:
```c
unsigned char *ctder = NULL;
int ctderlen, r;
ASN1_OBJECT* obj = OBJ_txt2obj("1.2.3.4", 1);
ctderlen = i2d_ASN1_OBJECT(obj, &ctder);
```
What is wrong with it?

### PR DESCRIPTION
**Pull request by levitte**
Since 0.9.7, all i2d_ functions were documented to allocate an output
buffer if the user didn't provide one, under these conditions (from
the 1.0.2 documentation):

    For OpenSSL 0.9.7 and later if B<*out> is B<NULL> memory will be
    allocated for a buffer and the encoded data written to it. In this
    case B<*out> is not incremented and it points to the start of the
    data just written.

i2d_ASN1_OBJECT was found not to do this, and would crash if a NULL
output buffer was provided.

Fixes #6914

### CODE REVIEW

**Comment by levitte** [comment]
Note that the 1.0.2 label is opportunistic, I haven't checked if it cherry-picks cleanly.  If not, I'll create a separate PR for 1.0.2...

**Review by mspncp — APPROVED** [review]
LGTM

**Review by mspncp — CHANGES_REQUESTED** [review]
Wait! One thing is not implemented:

> In this
case B<*out> is not incremented and it points to the start of the
data just written.

File: crypto/asn1/a_object.c [hunk 1]
```diff
@@ -30,6 +30,10 @@ int i2d_ASN1_OBJECT(const ASN1_OBJECT *a, unsigned char **pp)
     if (pp == NULL || objsize == -1)
         return objsize;
 
+    if (*pp == NULL
+        && (*pp = OPENSSL_malloc(objsize)) == NULL)
+            return 0;
```
Comment (mspncp) [inline crypto/asn1/a_object.c]: Maybe add `ASN1err(ASN1_F_I2D_ASN1_OBJECT, ERR_R_MALLOC_FAILURE);` ?

File: crypto/asn1/a_object.c [hunk 2]
```diff
@@ -30,12 +30,24 @@ int i2d_ASN1_OBJECT(const ASN1_OBJECT *a, unsigned char **pp)
...
     ASN1_put_object(&p, 0, a->length, V_ASN1_OBJECT, V_ASN1_UNIVERSAL);
     memcpy(p, a->data, a->length);
     p += a->length;
 
-    *pp = p;
+    /*
+     * If a new buffer was allocated, just return it back.
+     * If not, return the incremented buffer pointer.
+     */
+    *pp = allocated != NULL ? allocated : p;
```
Comment (mspncp) [inline crypto/asn1/a_object.c]: Nit: the `p += a->length`  statement is evaluated in vain in the case `allocated != NULL`. You could drop line 44 and replace line 50 by `*pp = allocated != NULL ? allocated : p + a->length;`

**Review by mspncp — APPROVED** [review]
The single build failure is unrelated (apt-get). Approved regardless whether the nit is addressed or not.

**Comment by levitte** [comment]
Merged.

master / 1.1.1:
cba024dc685d13dbcbd0577bed028ee6b295b56a i2d_ASN1_OBJECT(): allocate memory if the user didn't provide a buffer

1.1.0:
6114041540d8d1fecaf23a861788c3c742d3b467 i2d_ASN1_OBJECT(): allocate memory if the user didn't provide a buffer

1.0.2:
0971432f6f6d8b40d797133621809bd31eb7bf4e i2d_ASN1_OBJECT(): allocate memory if the user didn't provide a buffer

**Pull request closed without merging**